### DISCUSSION TITLE
iOS, System.Text.Json.JsonSerializer Exception coming.

### DISCUSSION THREAD

**Discussion by archergod** [thudugala/Plugin.LocalNotification#217] [Ideas]
Getting following Error (But it is not always, but 1 out of 2 attempt to debug my application give this error): 

`[0:] System.TypeInitializationException: The type initializer for 'System.Text.Json.JsonSerializer' threw an exception. ---> System.MissingMethodException: Method not found: int System.Text.Encodings.Web.TextEncoder.FindFirstCharacterToEncodeUtf8(System.ReadOnlySpan`1<byte>)
  at System.Text.Json.JsonEncodedText.EncodeHelper (System.ReadOnlySpan`1[T] utf8Value, System.Text.Encodings.Web.JavaScriptEncoder encoder) [0x00000] in <cb68364b029e41c3bf425990b94c94d9>:0 
  at System.Text.Json.JsonEncodedText.TranscodeAndEncode (System.ReadOnlySpan`1[T] value, System.Text.Encodings.Web.JavaScriptEncoder encoder) [0x00033] in <cb68364b029e41c3bf425990b94c94d9>:0 
  at System.Text.Json.JsonEncodedText.Encode (System.ReadOnlySpan`1[T] value, System.Text.Encodings.Web.JavaScriptEncoder encoder) [0x00014] in <cb68364b029e41c3bf425990b94c94d9>:0 
  at System.Text.Json.JsonEncodedText.Encode (System.String value, System.Text.Encodings.Web.JavaScriptEncoder encoder) [0x00014] in <cb68364b029e41c3bf425990b94c94d9>:0 
  at System.Text.Json.JsonSerializer..cctor () [0x00042] in <cb68364b029e41c3bf425990b94c94d9>:0 
   --- End of inner exception stack trace ---
  at Plugin.LocalNotification.NotificationCenter.GetRequestSerialize (Plugin.LocalNotification.NotificationRequest request) [0x0003e] in <763ab1a1f1b64389a6f1b291e8872c5b>:0 
  at Plugin.LocalNotification.Platform.iOS.NotificationServiceImpl.Show (Plugin.LocalNotification.NotificationRequest request) [0x000fe] in <763ab1a1f1b64389a6f1b291e8872c5b>:0 `

I found some topic that other libraries get this error, but I always get error on this Plugin for iOS, on exact those line. I try to include the Nuget package of `System.Text.Encoding.Web` that should add this function in code, but it is still not working and generating error. I see you use JsonSerializer Maybe you can include that library default (as requirement) Fix it ?

**Comment by thudugala** [comment]
Can you try adding this to the iOS App project:
```xml
<ItemGroup>
<PackageReference Include="System.Memory" Version="4.5.4">
<IncludeAssets>none</IncludeAssets>
</PackageReference>
<PackageReference Include="System.Buffers" Version="4.5.1">
<IncludeAssets>none</IncludeAssets>
</PackageReference>
</ItemGroup>
```

**Comment by archergod** [comment]
@thudugala  I try to add those 2 but no success. Even try to add to all projects .. still getting error.

**Comment by thudugala** [comment]
Can you please attach a sample project?

**Comment by archergod** [comment]
I don't have sample project yet. will see if I can create one later.

**Comment by archergod** [comment]
I didn't get chance to create sample project. But I did some more research and want to share, that some people suggest that .NETStandard2.0 might be a cause of this. Hiowever it is not true for me, as I didn't see .NETStandard2.0 in my project. Further more Some people blame it on Newtonsoft Json, which is there in my project, 

Second I try to use System.Memory 2.5.0 as one user say that it is only version that work for him, but didn't work for me. The error came when I try to create notification. So it is surely `Notification Plugin`. Sorry for not able to provide sample code. but my project use simple Sqlite PCL and newtonsoft to work with.

**Comment by jeremybirdwell80** [comment]
I'm am running into the same issue. I was able to strip down our app and get a sample together. I tried creating a new project to recreate it, but for some reason it didn't misbehave.
[DemoApp.zip](https://github.com/thudugala/Plugin.LocalNotification/files/7145658/DemoApp.zip)

**Comment by archergod** [comment]
https://docs.microsoft.com/en-us/dotnet/api/system.text.json?view=net-5.0&viewFallbackFrom=netframework-4.8  IT seems you are using System.Text.JSON in your code. which is not supported by older version on .NET framework. hence it is causing issues.

**Comment by archergod** [comment]
When I include the project directly (not as nuget package) it works. Also some research shows that error might be coming when JsonSerialization fails (that is why I add project directly) but I cannot reproduce it. 

However I still don't get notification.

**Comment by tranb3r** [comment]
Also happening on android, when app is launched in debug from visual studio 2022.
No error when launching the app on an emulator (not in debug).

**Comment by archergod** [comment]
The Error is due to new library of JSON by Microsoft, maybe if we use NewtonSoft's Library for Json operation it solve it?

**Comment by thudugala** [comment]
@archergod I was thinking maybe it should be pass to the user to serialize the string. What do you think?

**Comment by archergod** [comment]
@thudugala  I agree, then he can choose which library he wants.

BTW, my problem is solved for now, as I make a fork of project and use the source code directly in my as "Project in Solution". So I guess when you are making Nuget package it do not consider some extreme case. So you might leave it as is, and we suggest user with this problem to use Source code instead of Nuget Package.

**Comment by thudugala** [comment]
@archergod Please try v9.1.0-preview01

You can change the Serializer 

```cs
NotificationCenter.Serializer = new YourNotificationSerializer()

Public class YourNotificationSerializer : INotificationSerializer
{

}
```

**Comment by gaurakshay** [comment]
On my Google Pixel 3XL (updated to Android 12), when I try to schedule a notification using v9.1.0-preview01, I get the following message:

```
Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
```

I don't see a notification after I schedule it. I uninstalled my app before trying to deploy the app through visual studio again.

**Reply by tranb3r** [reply]
You should post the whole stacktrace.

**Reply by gaurakshay** [reply]
Unfortunately, no exception that bubbles up for me to get a stacktrace of. Here is the notification request along with the message that I can see in the output window.

```
[EGL_emulation] app_time_stats: avg=2485.97ms min=15.24ms max=39526.90ms count=16
Loaded assembly: /data/data/com.companyname.notif/files/.__override__/System.Numerics.dll [External]
Loaded assembly: /data/data/com.companyname.notif/files/.__override__/System.Data.dll [External]
[mpanyname.noti] Explicit concurrent copying GC freed 4220(1325KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 2682KB/5365KB, paused 333us total 6.881ms
[0:] Serialized Request [{"Android":{"AutoCancel":true,"ChannelId":"AkshayNotificationTest","Color":{"Argb":null,"ResourceName":null},"Group":null,"IconLargeName":{"ResourceName":null,"Type":"drawable"},"IconSmallName":{"ResourceName":null,"Type":"drawable"},"IsGroupSummary":false,"IsProgressBarIndeterminate":null,"LedColor":null,"Ongoing":false,"Priority":2,"ProgressBarMax":null,"ProgressBarProgress":null,"TimeoutAfter":null,"VibrationPattern":null,"VisibilityType":0},"BadgeNumber":0,"CategoryType":0,"Description":"Take an assessment","Image":{"Binary":null,"FilePath":null,"HasValue":false,"ResourceName":null},"iOS":{"HideForegroundAlert":false,"PlayForegroundSound":false},"NotificationId":1,"ReturningData":"This is the returning data.","Schedule":{"NotifyAutoCancelTime":null,"NotifyRepeatInterval":null,"NotifyTime":"2021-10-28T12:37:16.285683-05:00","RepeatType":0,"AndroidAllowedDelay":"00:01:00"},"Silent":false,"Sound":"","Subtitle":"","Title":"Test"}]
[CompatibilityChangeReporter] Compat change id reported: 160794467; UID 10149; state: ENABLED
[com.companyname.notif] com.companyname.notif: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
[com.companyname.notif] Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
[Choreographer] Skipped 50 frames!  The application may be doing too much work on its main thread.
[OpenGLRenderer] Davey! duration=866ms; Flags=0, FrameTimelineVsyncId=4994, IntendedVsync=830061445316, Vsync=830894778616, InputEventId=0, HandleInputStart=830909609000, AnimationStart=830909627100, PerformTraversalsStart=830910311300, DrawStart=830910418200, FrameDeadline=830094778648, FrameInterval=830909227100, FrameStartTime=16666666, SyncQueued=830910758300, SyncStart=830914033400, IssueDrawCommandsStart=830914104800, SwapBuffers=830914615000, FrameCompleted=830931226200, DequeueBufferDuration=15738500, QueueBufferDuration=368000, GpuCompleted=830929617100, SwapBuffersCompleted=830931226200, DisplayPresentTime=55448027792839, 
[EGL_emulation] app_time_stats: avg=5.53ms min=1.30ms max=20.96ms count=57
[EGL_emulation] app_time_stats: avg=136.97ms min=1.47ms max=2175.27ms count=17
Thread finished: <Thread Pool> #10
The thread 0xa has exited with code 0 (0x0).
```

**Comment by thudugala** [comment]
@archergod try Try 9.1.0-preview02

**Comment by archergod** [comment]
@thudugala I just recompile with new Preview 2 and it works like charm for me. I need to test further on real devices. But Serialization works great. (I use Newtonsoft) as I already have that in project.

**Reply by thudugala** [reply]
Preview 2 also use Newtonsoft
